lib/pos_transaction_register.go

```diff
@@ -0,0 +1,655 @@
+	maxFeeNanosPerKB uint64
+	// totalTxnsSizeBytes is the total size of all transactions in the FeeTimeBucket, in bytes.
+	totalTxnsSizeBytes uint64
+}
+
+func NewFeeTimeBucket(minFeeNanosPerKB uint64, maxFeeNanosPerKB uint64) *FeeTimeBucket {
+
+	txnsSet := treeset.NewWith(mempoolTxTimeOrderComparator)
+	return &FeeTimeBucket{
+		minFeeNanosPerKB:   minFeeNanosPerKB,
+		maxFeeNanosPerKB:   maxFeeNanosPerKB,
+		totalTxnsSizeBytes: 0,
+		txnsSet:            txnsSet,
+		txnMembership:      NewSet([]BlockHash{}),
+	}
+}
+
+// mempoolTxTimeOrderComparator is a comparator function for MempoolTx transactions stored inside a FeeTimeBucket. The comparator
+// orders the transactions by smallest timestamp. In case of a tie, transactions are ordered by greatest fee rate. Finally,
+// in case of another tie, transactions are ordered by their hash.
+func mempoolTxTimeOrderComparator(a, b interface{}) int {
+	aVal, aOk := a.(*MempoolTx)
+	bVal, bOk := b.(*MempoolTx)
+	if !aOk || !bOk {
+		glog.Error(CLog(Red, "mempoolTxTimeOrderComparator: Invalid types. This is BAD NEWS, we should never get here."))
+		return 0
+	}
+
+	if aVal.Added.UnixMicro() > bVal.Added.UnixMicro() {
+		return 1
+	} else if aVal.Added.UnixMicro() < bVal.Added.UnixMicro() {
+		return -1
+	} else if aVal.FeePerKB < bVal.FeePerKB {
+		return 1
+	} else if aVal.FeePerKB > bVal.FeePerKB {
+		return -1
+	}
+	// If the timestamps and fee rates are the same, we order by the transaction hash.
+	return bytes.Compare(aVal.Hash[:], bVal.Hash[:])
+}
+
+// AddTransaction adds a transaction to the FeeTimeBucket. It returns an error if the transaction is outside the
+// FeeTimeBucket's fee range, or if the transaction hash is nil.
+func (tb *FeeTimeBucket) AddTransaction(txn *MempoolTx) error {
+	tb.Lock()
+	defer tb.Unlock()
+
+	if txn == nil || txn.Hash == nil {
+		return fmt.Errorf("FeeTimeBucket.AddTransaction: Transaction or transaction hash is nil")
+	}
+
+	if tb.txnMembership.Includes(*txn.Hash) {
+		return nil
+	}
+
+	if tb.totalTxnsSizeBytes > math.MaxUint64-txn.TxSizeBytes {
+		return fmt.Errorf("FeeTimeBucket.AddTransaction: Transaction size %d would overflow bucket size %d",
+			txn.TxSizeBytes, tb.totalTxnsSizeBytes)
+	}
+
+	if tb.minFeeNanosPerKB > txn.FeePerKB || tb.maxFeeNanosPerKB < txn.FeePerKB {
+		return fmt.Errorf("FeeTimeBucket.AddTransaction: Transaction fee %d outside of bucket range [%d, %d]",
+			txn.FeePerKB, tb.minFeeNanosPerKB, tb.maxFeeNanosPerKB)
+	}
+
+	tb.txnsSet.Add(txn)
+	tb.txnMembership.Add(*txn.Hash)
+	tb.totalTxnsSizeBytes += txn.TxSizeBytes
+	return nil
+}
+
+// RemoveTransaction removes a transaction from the FeeTimeBucket.
+func (tb *FeeTimeBucket) RemoveTransaction(txn *MempoolTx) {
+	tb.Lock()
+	defer tb.Unlock()
+
+	if txn == nil || txn.Hash == nil {
+		return
+	}
+
+	if !tb.txnMembership.Includes(*txn.Hash) {
+		return
+	}
+
+	tb.txnsSet.Remove(txn)
+	tb.txnMembership.Remove(*txn.Hash)
```

Comment:
Why do we track txns with two sets? Why not just `map[BlockHash]*MempoolTx`

Reply:
We need the `txnsSet` for Fee-Time ordering of transactions. When it comes to `txnMembership`, I actually started with map but we decied to use the Set during review. The map and the Set are very similar. In fact Set is just a wrapper around a map. I think it's slightly preferable to use the Set since it comes with more functionality.